File: Renaissance.AspNet/frontend/Renaissance.Web/Models/ReferralModels.cs
namespace Renaissance.Web.Models;

public enum ReferralStatus
{
    Pending = 1,
    InProgress = 2,
    Completed = 3,
    Cancelled = 4
}

public enum ReferralPriority
{
    Emergency = 1,
    Urgent = 2,
    Routine = 3
}

public class CreateReferralsRequest
{
    public Guid PatientId { get; set; }
    public AppModule SourceModule { get; set; }
    public Guid? SourceRecordId { get; set; }
    public List<AppModule> TargetModules { get; set; } = [];
    public string? Notes { get; set; }
    public ReferralPriority Priority { get; set; } = ReferralPriority.Routine;
}

public class ReassignReferralRequest
{
    public AppModule TargetModule { get; set; }
    public string? Notes { get; set; }
}

public class ReferralQueueItem
{
    public Guid Id { get; set; }
    public Guid PatientId { get; set; }
    public string PatientFullName { get; set; } = string.Empty;
    public string ClientNumber { get; set; } = string.Empty;
    public int? PatientAge { get; set; }
    public string? PatientAgeUnit { get; set; }
    public string? PatientSex { get; set; }
    public AppModule SourceModule { get; set; }
    public AppModule TargetModule { get; set; }
    public ReferralStatus Status { get; set; }
    public ReferralPriority Priority { get; set; }
    public string? Notes { get; set; }
    public string? ReferredBy { get; set; }
    public string? AttendedBy { get; set; }
    public string? CompletedBy { get; set; }
    public DateTime? CreatedDate { get; set; }
    public DateTime? AttendedDate { get; set; }
    public DateTime? CompletedDate { get; set; }
    public int QueuePosition { get; set; }
    public int WaitMinutes { get; set; }
    public bool IsSlaBreached { get; set; }
}

public class ReferralNotification
{
    public Guid ReferralId { get; set; }
    public Guid PatientId { get; set; }
    public string PatientFullName { get; set; } = string.Empty;
    public string ClientNumber { get; set; } = string.Empty;
    public AppModule SourceModule { get; set; }
    public AppModule TargetModule { get; set; }
    public ReferralPriority Priority { get; set; }
    public string? Notes { get; set; }
    public DateTime? CreatedDate { get; set; }
}

public class ReferralInboxItem
{
    public Guid ReferralId { get; set; }
    public Guid PatientId { get; set; }
    public string PatientFullName { get; set; } = string.Empty;
    public string ClientNumber { get; set; } = string.Empty;
    public AppModule SourceModule { get; set; }
    public AppModule TargetModule { get; set; }
    public ReferralStatus Status { get; set; }
    public ReferralPriority Priority { get; set; }
    public string? Notes { get; set; }
    public DateTime? CreatedDate { get; set; }
    public int WaitMinutes { get; set; }
    public bool IsRead { get; set; }
}

public class ModuleReferralCount
{
    public AppModule Module { get; set; }
    public int PendingCount { get; set; }
}

public class PatientJourneyStep
{
    public Guid ReferralId { get; set; }
    public AppModule TargetModule { get; set; }
    public AppModule SourceModule { get; set; }
    public ReferralStatus Status { get; set; }
    public ReferralPriority Priority { get; set; }
    public string? ReferredBy { get; set; }
    public string? AttendedBy { get; set; }
    public string? CompletedBy { get; set; }
    public DateTime? CreatedDate { get; set; }
    public DateTime? AttendedDate { get; set; }
    public DateTime? CompletedDate { get; set; }
    public string? Notes { get; set; }
}

public class CreateReferralsResult
{
    public List<ReferralQueueItem> Created { get; set; } = [];
    public List<AppModule> SkippedDuplicates { get; set; } = [];
}

File: Renaissance.AspNet/frontend/Renaissance.Angular/src/app/core/models/referral.models.ts
import { AppModule } from './auth.models';

export enum ReferralStatus {
    Pending = 1,
    InProgress = 2,
    Completed = 3,
    Cancelled = 4
}

export enum ReferralPriority {
    Emergency = 1,
    Urgent = 2,
    Routine = 3
}

export interface CreateReferralsRequest {
    patientId: string;
    sourceModule: AppModule;
    sourceRecordId?: string;
    targetModules: AppModule[];
    notes?: string;
    priority: ReferralPriority;
}

export interface ReassignReferralRequest {
    targetModule: AppModule;
    notes?: string;
}

export interface ReferralQueueItem {
    id: string;
    patientId: string;
    patientFullName: string;
    clientNumber: string;
    patientAge?: number;
    patientAgeUnit?: string;
    patientSex?: string;
    sourceModule: AppModule;
    targetModule: AppModule;
    status: ReferralStatus;
    priority: ReferralPriority;
    notes?: string;
    referredBy?: string;
    attendedBy?: string;
    completedBy?: string;
    createdDate?: string;
    attendedDate?: string;
    completedDate?: string;
    queuePosition: number;
    waitMinutes: number;
    isSlaBreached: boolean;
}

export interface ReferralNotification {
    referralId: string;
    patientId: string;
    patientFullName: string;
    clientNumber: string;
    sourceModule: AppModule;
    targetModule: AppModule;
    priority: ReferralPriority;
    notes?: string;
    createdDate?: string;
}

export interface ReferralInboxItem {
    referralId: string;
    patientId: string;
    patientFullName: string;
    clientNumber: string;
    sourceModule: AppModule;
    targetModule: AppModule;
    status: ReferralStatus;
    priority: ReferralPriority;
    notes?: string;
    createdDate?: string;
    waitMinutes: number;
    isRead: boolean;
}

export interface ModuleReferralCount {
    module: AppModule;
    pendingCount: number;
}

export interface PatientJourneyStep {
    referralId: string;
    targetModule: AppModule;
    sourceModule: AppModule;
    status: ReferralStatus;
    priority: ReferralPriority;
    referredBy?: string;
    attendedBy?: string;
    completedBy?: string;
    createdDate?: string;
    attendedDate?: string;
    completedDate?: string;
    notes?: string;
}

export interface CreateReferralsResult {
    created: ReferralQueueItem[];
    skippedDuplicates: AppModule[];
}
